mfs/system.go

```diff
@@ -0,0 +1,237 @@
+// package mfs implements an in memory model of a mutable ipfs filesystem.
+//
+// It consists of four main structs:
+// 1) The Filesystem
+//        The filesystem serves as a container and entry point for various mfs filesystems
+// 2) Root
+//        Root represents an individual filesystem mounted within the mfs system as a whole
+// 3) Directories
+// 4) Files
+package mfs
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	eventlog "github.com/ipfs/go-ipfs/thirdparty/eventlog"
+)
+
+var ErrNotExist = errors.New("no such rootfs")
+
+var log = eventlog.Logger("mfs")
+
+var ErrIsDirectory = errors.New("error: is a directory")
+
+type childCloser interface {
+	closeChild(string, *dag.Node) error
+}
+
+type NodeType int
+
+const (
+	TFile NodeType = iota
+	TDir
+)
+
+// FSNode represents any node (directory, root, or file) in the ipns filesystem
+type FSNode interface {
+	GetNode() (*dag.Node, error)
+	Type() NodeType
+	Lock()
+	Unlock()
+}
+
+// Root represents the root of a filesystem tree pointed to by a given keypair
+type Root struct {
+	// node is the merkledag node pointed to by this keypair
+	node *dag.Node
+
+	// val represents the node pointed to by this key. It can either be a File or a Directory
+	val FSNode
+
+	repub *Republisher
+
+	dserv dag.DAGService
+
+	Type string
+}
+
+type PubFunc func(context.Context, key.Key) error
+
+// newRoot creates a new Root for the given key, and starts up a republisher routine
+// for it
+func NewRoot(parent context.Context, ds dag.DAGService, node *dag.Node, pf PubFunc) (*Root, error) {
+	ndk, err := node.Key()
+	if err != nil {
+		return nil, err
+	}
+
+	root := &Root{
+		node:  node,
+		repub: NewRepublisher(parent, pf, time.Millisecond*300, time.Second*3),
```

Comment:
i'm not opposed to that. But i also couldnt think of any applications that really cared about every single intermediate state. (vim wrote a .swp file, vim changed the swap file because you pressed a key, vim changed the swap file because... ad nauseum)